src/main/java/com/artipie/rpm/misc/RpmByDigestCopy.java

```diff
@@ -0,0 +1,141 @@
+    private final ContentDigest.Digests algorithm;
+
+    /**
+     * Ctor.
+     * @param from Storage to copy from
+     * @param key Key to copy from
+     * @param digests Content digests to exclude
+     * @param algorithm Digest algorithm
+     * @checkstyle ParameterNumberCheck (5 lines)
+     */
+    public RpmByDigestCopy(
+        final Storage from, final Key key, final List<String> digests,
+        final ContentDigest.Digests algorithm
+    ) {
+        this.from = from;
+        this.digests = digests;
+        this.key = key;
+        this.algorithm = algorithm;
+    }
+
+    /**
+     * Ctor.
+     * @param from Storage to copy from
+     * @param key Key to copy from
+     * @param digests Content digests to exclude
+     */
+    public RpmByDigestCopy(final Storage from, final Key key, final List<String> digests) {
+        this(from, key, digests, ContentDigest.Digests.SHA256);
+    }
+
+    /**
+     * Copy rpm to destination storage filtering by digest.
+     * @param dest Destination
+     * @return Completable copy operation
+     */
+    Completable copy(final Storage dest) {
+        return SingleInterop.fromFuture(this.from.list(this.key))
+            .flatMapPublisher(Flowable::fromIterable)
+            .filter(item -> item.string().endsWith(".rpm"))
+            .flatMapCompletable(
+                rpm -> Completable.fromFuture(
+                    this.from.value(rpm).thenCompose(
+                        content -> RpmByDigestCopy.bytes(content).thenCompose(
+                            source -> new ContentDigest(new Content.From(source), this.algorithm)
+                                .hex().thenCompose(
+                                    hex -> {
+                                        final String file = Paths.get(rpm.string()).getFileName()
+                                            .toString();
+                                        final CompletableFuture<Void> res;
+                                        if (this.digests.contains(hex)) {
+                                            res = CompletableFuture.allOf();
+                                        } else {
+                                            res = dest.save(
+                                                new Key.From(file), new Content.From(source)
+                                            );
+                                        }
+                                        return res;
+                                    }
+                                )
+                            )
+                    )
+                )
+            );
+    }
+
+    /**
+     * Bytes from content.
+     * @param content Content
+     * @return Bytes
+     */
+    private static CompletionStage<byte[]> bytes(final Content content) {
+        return new Concatenation(content)
+            .single()
+            .map(buf -> new Remaining(buf, true))
+            .map(Remaining::bytes)
+            .to(SingleInterop.get());
+    }
```

Comment:
@olenagerasimova could be now replaced with `PublisherAs` usage

Reply:
@olegmoz I have separate issue to update asto and `PublisherAs` as several places should be fixed, I think it would be better to leave it as is for now and fix when updating asto